packages/devtools_app/lib/src/performance/performance_controller.dart

```diff
@@ -237,10 +224,33 @@ class PerformanceController extends DisposableController
         : PerformanceData();
   }
 
-  FutureOr<void> processAvailableEvents() async {
-    if (data == null) {
-      await _initData();
+  Future<void> _pullTraceEventsFromVmTimeline({
+    bool shouldPrimeThreadIds = false,
+  }) async {
+    final currentVmTime = await serviceManager.service.getVMTimelineMicros();
+    debugTraceEventLog(
+      'pulling trace events from '
+      '[$_nextPollStartMicros - ${currentVmTime.timestamp}]',
+    );
+    final timeline = await serviceManager.service.getVMTimeline(
+      timeOriginMicros: _nextPollStartMicros,
+      timeExtentMicros: currentVmTime.timestamp - _nextPollStartMicros,
+    );
+    _nextPollStartMicros = currentVmTime.timestamp + 1;
+
+    // TODO(kenz): move this priming logic into the loop below.
+    if (shouldPrimeThreadIds) primeThreadIds(timeline);
+    for (final event in timeline.traceEvents) {
+      final eventWrapper = TraceEventWrapper(
+        TraceEvent(event.json),
+        DateTime.now().millisecondsSinceEpoch,
+      );
+      allTraceEvents.add(eventWrapper);
+      debugTraceEventLog(eventWrapper.event.json.toString());
```

Comment:
Can you remove this debugTraceEventLog or pass the raw json object instead of calling toString? This code will call the toString on the JSON even in release mode which could be somewhat expensive.

Reply:
changing this method to be a callback so that the toString() won't get called unless the callback is called